verification_service/main.py

```diff
@@ -0,0 +1,309 @@
+            # get it from their GitLab user attributes.
+            if not user_email_address and user_id_with_max_access is not None:
+                response = requests.get(
+                    f"{self.gitlab_url}/api/v4/users/{user_id_with_max_access}",
+                    headers={"PRIVATE-TOKEN": f"{self.gitlab_access_token}"},
+                )
+                user = response.json()
+                user_email_address = user.get("email")
+
+        # If an email address is still not located and the event type
+        # is not snippet_check, log the situation and return.
+        if user_email_address is None and event_type != "snippet_check":
+            logging.debug(
+                f"Unable to get user email address for event type: {event_type}"
+            )
+            return
+
+        # If an email address is located, check if the user or their
+        # email domain is verified.
+        elif user_email_address is not None and event_type != "snippet_check":
+            user_verified = check_domain_verification(
+                user_email_address, self.verified_domains_file
+            )
+
+            if not user_verified:
+                user_verified = check_user_verification(
+                    user_email_address, self.verified_users_file
+                )
+
+            logging.info(
+                f"{user_email_address}, domain verification: {user_verified}"
+            )
+
+        # Check if the event type is snippet_check.
+        elif event_type == "snippet_check":
+            logging.info(
+                "Snippet check event received."
+                "Individual snippet verification will be done at a later point."
+                "Passing event to the next service."
+            )
+            snippet_check_events_total.inc()
+
+        # If the user or their email domain is not verified, push the
+        # event to the next service. If the user or their email domain
+        # is verified, log the situation and return.
+        if not user_verified:
+            verification_failures_total.inc()
+            self.send_to_queue(event_type, data, prefix="verification")
+            logging.debug(
+                f"Pushed event to queue: verification_{event_type}"
+            )
+        else:
+            verified_events_total.inc()
+            logging.info(
+                f"Verified event type {event_type} for user {user_email_address}"
+            )
+
+
+def process_events(
+    redis_conn=None,
+    verified_users_file=None,
+    verified_domains_file=None,
+    gitlab_url=None,
+    gitlab_access_token=None,
+):
+    processor = VerificationEventProcessor(
+        "event",
+        redis_conn=redis_conn,
+        verified_users_file=verified_users_file,
+        verified_domains_file=verified_domains_file,
+        gitlab_url=gitlab_url,
+        gitlab_access_token=gitlab_access_token,
+    )
+
+    while True:
+        processor.retrieve_event()
+        sleep(1)
```

Comment:
At most here you can handle 1 event per second. Same as my other comments - this is a great place for asyncio.

Reply:
See above comment regarding GitLab API